### DISCUSSION TITLE
Inquiry about the method to increase the optimisation speed

### DISCUSSION THREAD

**Discussion by HaoyuFang0909** [spine-tools/SpineOpt.jl#727] [Support]
Hi @DillonJ ,

Based on a SPINE model of the UK electricity system, I would like to run the optimization for a whole year, but I cannot run it at 1-hour resolution as it would take quite a lot of time. (I ran it at 12-hour resolution and it takes 10 hours) I wonder if there is any method to speed up the optimisation? 

Thanks

**Comment by jkiviluo** [comment]
Hello. Couple of questions to narrow the problem:

- Which solver are you using? 
- Is the problem LP or MIP? If the latter, what decision variables are binary/integer and how many of those there are?
- Are you running it on laptop or something more powerful? 
- Are you running investments or just scheduling/dispatch? If the latter, are you using rolling optimization or single shot?
- How many units and nodes you have? (or how many nonzero variables your problem has, either the 12-hour or the 1-hour version)?

**Comment by HaoyuFang0909** [comment]
[Generator_databasev2023_06_02.zip](https://github.com/spine-tools/SpineOpt.jl/files/12303589/Generator_databasev2023_06_02.zip)

Thanks for your timely response! @jkiviluo 

I'm using HiGS as the LP solver and Cbc as the MIP solver. I'm running it on my laptop. It is running scheduling including investments. There are nearly 900 elements in the model, and all data is 1-hour version.

I would appreciate it if you could check the model and give some advice.

Thanks!

**Comment by DillonJ** [comment]
Hi @HaoyuFang0909 why not use HiGHS as the MIP solver? It's a lot more performant than CBC. But this might not solve your issue.

Can your model use a rolling optimisation? What is the longest duration storage in the model? Does it have investments?

**Comment by jkiviluo** [comment]
At a quick glance, it doesn't look like a big model and you're not running investments. What's probably making it slow is that you have binary start-up variables for most of your units and long minimum uptime and downtime requirements for some of them. Those can make the model slow, since they create difficult constraints for the solver. You could try to remove the minimum down and uptime requirements from most units (typically that's not going to affect your results very much), just leave them on units like nuclear for which they are more serious. You could also create an alternative scenario where the unit online decisions are linear to see how fast you can get it that way.

Also, CBC is very slow (which you would be using in this integer model). You could try HiGHS. It should be faster on MIP problems too. However, it will still be much slower than commercial solvers (in MIP).

Jody's suggestion to use rolling optimization is also very good if you want to keep the level of detail you have for the thermal units.

**Comment by HaoyuFang0909** [comment]
Thanks for your helpful advice! @DillonJ @jkiviluo 

Sorry I don't know much about the performance of these solvers. 

For rolling optimization, how should I set `roll_forward` to consider the power generation throughout the year to achieve the goal of seasonal storage? If I set `roll_forward` as 1M only, the model will only consider the generation in one month.

**Comment by HaoyuFang0909** [comment]
Also, I am not sure that you said I was not running investments. I have added the temporal block and parameters of Electrolysis and Fuel Cells for investment. Is there something wrong with my investment part?

**Comment by HaoyuFang0909** [comment]
In the result the investment is indeed not being used, I suspect the investment price is set too high, or there is indeed a problem?

**Comment by HaoyuFang0909** [comment]
Sorry to bother you again. When I set `HiGHS` as the MIP solver, an error appeared as follows: 

ERROR: LoadError: MathOptInterface.UnsupportedAttribute{MathOptInterface.RawParameter}: Attribute MathOptInterface.RawParameter("mip_rel_gap") is not supported by the model.
Stacktrace:
  [1] set(model::HiGHS.Optimizer, param::MathOptInterface.RawParameter, value::Float64)
    @ HiGHS C:\Users\10926\.julia\packages\HiGHS\YP79q\src\MOI_wrapper.jl:503
  [2] _instantiate_and_check(optimizer_constructor::MathOptInterface.OptimizerWithAttributes)
    @ MathOptInterface C:\Users\10926\.julia\packages\MathOptInterface\YDdD3\src\instantiate.jl:88
  [3] instantiate(optimizer_constructor::MathOptInterface.OptimizerWithAttributes; with_bridge_type::Type{Float64}, with_names::Bool)
    @ MathOptInterface C:\Users\10926\.julia\packages\MathOptInterface\YDdD3\src\instantiate.jl:120
  [4] set_optimizer(model::JuMP.Model, optimizer_constructor::MathOptInterface.OptimizerWithAttributes; bridge_constraints::Bool)
    @ JuMP C:\Users\10926\.julia\packages\JuMP\qhoVb\src\optimizer_interface.jl:66
  [5] set_optimizer
    @ C:\Users\10926\.julia\packages\JuMP\qhoVb\src\optimizer_interface.jl:59 [inlined]
  [6] #Model#23
    @ C:\Users\10926\.julia\packages\JuMP\qhoVb\src\JuMP.jl:235 [inlined]
  [7] Model
    @ C:\Users\10926\.julia\packages\JuMP\qhoVb\src\JuMP.jl:232 [inlined]
  [8] _do_create_model(mip_solver::MathOptInterface.OptimizerWithAttributes, use_direct_model::Bool)
    @ SpineOpt C:\Users\10926\.julia\packages\SpineOpt\Flpc7\src\run_spineopt.jl:397
  [9] #invokelatest#2
    @ .\essentials.jl:816 [inlined]
 [10] invokelatest
    @ .\essentials.jl:813 [inlined]
 [11] create_model(model_type::Symbol, mip_solver::Nothing, lp_solver::Nothing, use_direct_model::Bool, is_subproblem::Bool)
    @ SpineOpt C:\Users\10926\.julia\packages\SpineOpt\Flpc7\src\run_spineopt.jl:321
 [12] rerun_spineopt(url_out::String; mip_solver::Nothing, lp_solver::Nothing, add_user_variables::Function, add_constraints::Function, update_constraints::Function, log_level::Int64, optimize::Bool, update_names::Bool, alternative::String, write_as_roll::Int64, resume_file_path::Nothing, use_direct_model::Bool, alternative_objective::SpineOpt.var"#90#98")
    @ SpineOpt C:\Users\10926\.julia\packages\SpineOpt\Flpc7\src\run_spineopt.jl:278
 [13] _run_spineopt(url_in::String, url_out::String; upgrade::Bool, mip_solver::Nothing, lp_solver::Nothing, add_user_variables::Function, add_constraints::Function, update_constraints::Function, log_level::Int64, optimize::Bool, update_names::Bool, alternative::String, write_as_roll::Int64, use_direct_model::Bool, filters::Dict{String, String}, resume_file_path::Nothing)
    @ SpineOpt C:\Users\10926\.julia\packages\SpineOpt\Flpc7\src\run_spineopt.jl:194
 [14] _run_spineopt
    @ C:\Users\10926\.julia\packages\SpineOpt\Flpc7\src\run_spineopt.jl:173 [inlined]
 [15] #run_spineopt#43
    @ C:\Users\10926\.julia\packages\SpineOpt\Flpc7\src\run_spineopt.jl:102 [inlined]
 [16] run_spineopt(url_in::String, url_out::String)
    @ SpineOpt C:\Users\10926\.julia\packages\SpineOpt\Flpc7\src\run_spineopt.jl:82
 [17] top-level scope
    @ C:\Users\10926\.spinetoolbox\work\run_spineopt__53719c9f9435475f9d5cca0ac8db1d91__toolbox\Default:3
 [18] include(fname::String)
    @ Base.MainInclude .\client.jl:478
 [19] top-level scope
    @ none:1
in expression starting at C:\Users\10926\.spinetoolbox\work\run_spineopt__53719c9f9435475f9d5cca0ac8db1d91__toolbox\Default:3

**Comment by HaoyuFang0909** [comment]
Thanks! I have solved this problem by specifying HiGHS in `Run_SpineOpt`
